### PR TITLE
Add category model

### PR DESCRIPTION
**Pull request by layzimayner**

### CODE REVIEW

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "categories management", description = "Endpoints for management categories")
+@RestController
+@RequestMapping("/categories")
+@Validated
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/controller/CategoryController.java:31]: How does this annotation work? Do you use it?

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+@RequiredArgsConstructor
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/controller/CategoryController.java:39]: ResponseStatus.CREATED

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    @Operation(summary = "Create new category", description = "Add new category to DB")
+    public CategoryDto createCategory(@RequestBody @Valid CreateCategoryRequestDto requestDto) {
+        return categoryService.save(requestDto);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get list of all available categories")
+    public List<CategoryDto> getAll(Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/controller/CategoryController.java:47]: better return Page (optional)

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @Operation(summary = "Get category", description = "Get category by id")
+    public CategoryDto getCategoryById(Long id) {
+        return categoryService.getById(id);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Update category", description = "Update category by id")
+    public CategoryDto updateCategory(
+            @PathVariable @Valid Long id,
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/controller/CategoryController.java]: How are you validating an ID? You can use Positive

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete category", description = "Delete category by id")
+    public void deleteCategory(Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/by-categoryId")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/controller/CategoryController.java]: following task description `/api/categories/{id}/books`

File: src/main/java/com/example/demo/dto/book/CreateBookRequestDto.java [hunk 6]
```diff
@@ -20,4 +21,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoriesIds;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/dto/book/CreateBookRequestDto.java:26]: Can a book exist without a Category? Let’s  add NotEmpty validator

File: src/main/java/com/example/demo/model/Book.java [hunk 7]
```diff
@@ -34,4 +41,12 @@ public class Book {
     private String coverImage;
     @Column(nullable = false, columnDefinition = "TINYINT(1) DEFAULT 0")
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/model/Book.java]:
```suggestion
    @ManyToMany
```
It's default fetch type

File: src/main/java/com/example/demo/model/User.java [hunk 8]
```diff
@@ -37,12 +40,13 @@ public class User implements UserDetails {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
-    @ManyToMany
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/model/User.java]:
```suggestion
    @ManyToMany
```
Let’s  use join fetch or EntityGraph

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+              - column:
+                  name: id
+                  type: BIGINT
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: VARCHAR(50)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml:18]: validate request dto length

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                    unique: true
+              - column:
+                  name: description
+                  type: VARCHAR(255)
+              - column:
+                  name: is_deleted
+                  type: TINYINT(1)
+                  defaultValue: 0
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml]: add one empty line at the end of the file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get list of all available categories")
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get category", description = "Get category by id")
+    public CategoryDto getCategoryById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/controller/CategoryController.java]:
```suggestion
    public CategoryDto getCategoryById(@PathVariable @Positive Long id) {
```

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        return categoryService.save(requestDto);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get list of all available categories")
+    public List<CategoryDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @GetMapping
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/controller/CategoryController.java]:
```suggestion
    @GetMapping("/{id}")
```

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            @PathVariable @Positive Long id,
+            @RequestBody @Valid CreateCategoryRequestDto categoryDto) {
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete category", description = "Delete category by id")
+    public void deleteCategory(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/controller/CategoryController.java]:
```suggestion
    public void deleteCategory(@PathVariable @Positive Long id)) {
```

File: src/main/java/com/example/demo/controller/CategoryController.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/{categoryId}/books")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@RequestParam Long categoryId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/controller/CategoryController.java]:
```suggestion
    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable @Positive Long categoryId) {
```

File: src/main/java/com/example/demo/dto/book/CreateBookRequestDto.java [hunk 15]
```diff
@@ -20,4 +21,6 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/dto/book/CreateBookRequestDto.java]:
```suggestion
    @NotEmpty
```

File: src/main/java/com/example/demo/model/Book.java [hunk 16]
```diff
@@ -22,16 +29,33 @@ public class Book {
...
     @Column(nullable = false)
     private BigDecimal price;
+
     private String description;
     private String coverImage;
+
     @Column(nullable = false, columnDefinition = "TINYINT(1) DEFAULT 0")
     private boolean isDeleted = false;
+
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]:
```suggestion
```

File: src/main/java/com/example/demo/model/Book.java [hunk 17]
```diff
@@ -22,16 +29,33 @@ public class Book {
...
+    @ManyToMany
+    @JoinTable(
+            name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id"))
+
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private Set<Category> categories;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]:
```suggestion
    private Set<Category> categories = new HashSet<>();
```

File: src/main/java/com/example/demo/model/User.java (hunk 8 above)
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/User.java]: don't use Eager fetchtype

```suggestion
    @ManyToMany
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments, 
build failed, fix it

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 18]
```diff
@@ -13,3 +13,7 @@ databaseChangeLog:
       file: db/changelog/changes/06-insert-users.yaml
   - include:
       file: db/changelog/changes/07-add-is_deleted-column-to-users.yaml
+  - include:
+      file: db/changelog/changes/08-create-categories-table.yaml
+  - include:
+      file: db/changelog/changes/09-create-books-categories-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: add one empty line

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**